<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.10.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>AREA: Prevent calling &lt;tt&gt;expect&lt;/tt&gt; conditionally (&lt;tt&gt;no-conditional-expect&lt;/tt&gt;)</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<script type="text/javascript" src="clipboard.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">AREA
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.10.0 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

</div><!-- top -->
<div><div class="header">
  <div class="headertitle"><div class="title">Prevent calling &lt;tt&gt;expect&lt;/tt&gt; conditionally (&lt;tt&gt;no-conditional-expect&lt;/tt&gt;)</div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a class="anchor" id="autotoc_md10772"></a> This rule prevents the use of <code>expect</code> in conditional blocks, such as <code>if</code>s &amp; <code>catch</code>s.</p>
<p>This includes using <code>expect</code> in callbacks to functions named <code>catch</code>, which are assumed to be promises.</p>
<h1><a class="anchor" id="autotoc_md10773"></a>
Rule Details</h1>
<p>Jest only considers a test to have failed if it throws an error, meaning if calls to assertion functions like <code>expect</code> occur in conditional code such as a <code>catch</code> statement, tests can end up passing but not actually test anything.</p>
<p>Additionally, conditionals tend to make tests more brittle and complex, as they increase the amount of mental thinking needed to understand what is actually being tested.</p>
<p>While <code>expect.assertions</code> &amp; <code>expect.hasAssertions</code> can help prevent tests from silently being skipped, when combined with conditionals they typically result in even more complexity being introduced.</p>
<p>The following patterns are warnings:</p>
<div class="fragment"><div class="line">it(&#39;foo&#39;, () =&gt; {</div>
<div class="line">  doTest &amp;&amp; expect(1).toBe(2);</div>
<div class="line">});</div>
<div class="line"> </div>
<div class="line">it(&#39;bar&#39;, () =&gt; {</div>
<div class="line">  if (!skipTest) {</div>
<div class="line">    expect(1).toEqual(2);</div>
<div class="line">  }</div>
<div class="line">});</div>
<div class="line"> </div>
<div class="line">it(&#39;baz&#39;, async () =&gt; {</div>
<div class="line">  try {</div>
<div class="line">    await foo();</div>
<div class="line">  } catch (err) {</div>
<div class="line">    expect(err).toMatchObject({ code: &#39;MODULE_NOT_FOUND&#39; });</div>
<div class="line">  }</div>
<div class="line">});</div>
<div class="line"> </div>
<div class="line">it(&#39;throws an error&#39;, async () =&gt; {</div>
<div class="line">  await foo().catch(error =&gt; expect(error).toBeInstanceOf(error));</div>
<div class="line">});</div>
</div><!-- fragment --><p>The following patterns are not warnings:</p>
<div class="fragment"><div class="line">it(&#39;foo&#39;, () =&gt; {</div>
<div class="line">  expect(!value).toBe(false);</div>
<div class="line">});</div>
<div class="line"> </div>
<div class="line">function getValue() {</div>
<div class="line">  if (process.env.FAIL) {</div>
<div class="line">    return 1;</div>
<div class="line">  }</div>
<div class="line"> </div>
<div class="line">  return 2;</div>
<div class="line">}</div>
<div class="line"> </div>
<div class="line">it(&#39;foo&#39;, () =&gt; {</div>
<div class="line">  expect(getValue()).toBe(2);</div>
<div class="line">});</div>
<div class="line"> </div>
<div class="line">it(&#39;validates the request&#39;, () =&gt; {</div>
<div class="line">  try {</div>
<div class="line">    processRequest(request);</div>
<div class="line">  } catch {</div>
<div class="line">    // ignore errors</div>
<div class="line">  } finally {</div>
<div class="line">    expect(validRequest).toHaveBeenCalledWith(request);</div>
<div class="line">  }</div>
<div class="line">});</div>
<div class="line"> </div>
<div class="line">it(&#39;throws an error&#39;, async () =&gt; {</div>
<div class="line">  await expect(foo).rejects.toThrow(Error);</div>
<div class="line">});</div>
</div><!-- fragment --><h2><a class="anchor" id="autotoc_md10774"></a>
How to catch a thrown error for testing without violating this rule</h2>
<p>A common situation that comes up with this rule is when wanting to test properties on a thrown error, as Jest's <code>toThrow</code> matcher only checks the <code>message</code> property.</p>
<p>Most people write something like this:</p>
<div class="fragment"><div class="line">describe(&#39;when the http request fails&#39;, () =&gt; {</div>
<div class="line">  it(&#39;includes the status code in the error&#39;, async () =&gt; {</div>
<div class="line">    try {</div>
<div class="line">      await makeRequest(url);</div>
<div class="line">    } catch (error) {</div>
<div class="line">      expect(error).toHaveProperty(&#39;statusCode&#39;, 404);</div>
<div class="line">    }</div>
<div class="line">  });</div>
<div class="line">});</div>
</div><!-- fragment --><p>As stated above, the problem with this is that if <code>makeRequest()</code> doesn't throw the test will still pass as if the <code>expect</code> had been called.</p>
<p>While you can use <code>expect.assertions</code> &amp; <code>expect.hasAssertions</code> for these situations, they only work with <code>expect</code>.</p>
<p>A better way to handle this situation is to introduce a wrapper to handle the catching, and otherwise returns a specific "no error thrown" error if nothing is thrown by the wrapped function:</p>
<div class="fragment"><div class="line">class NoErrorThrownError extends Error {}</div>
<div class="line"> </div>
<div class="line">const getError = async &lt;TError&gt;(call: () =&gt; unknown): Promise&lt;TError&gt; =&gt; {</div>
<div class="line">  try {</div>
<div class="line">    await call();</div>
<div class="line"> </div>
<div class="line">    throw new NoErrorThrownError();</div>
<div class="line">  } catch (error: unknown) {</div>
<div class="line">    return error as TError;</div>
<div class="line">  }</div>
<div class="line">};</div>
<div class="line"> </div>
<div class="line">describe(&#39;when the http request fails&#39;, () =&gt; {</div>
<div class="line">  it(&#39;includes the status code in the error&#39;, async () =&gt; {</div>
<div class="line">    const error = await getError(async () =&gt; makeRequest(url));</div>
<div class="line"> </div>
<div class="line">    // check that the returned error wasn&#39;t that no error was thrown</div>
<div class="line">    expect(error).not.toBeInstanceOf(NoErrorThrownError);</div>
<div class="line">    expect(error).toHaveProperty(&#39;statusCode&#39;, 404);</div>
<div class="line">  });</div>
<div class="line">});</div>
</div><!-- fragment --> </div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.10.0
</small></address>
</body>
</html>
